### PR TITLE
DAOS-9208 vos: add logic offset anchor for recx enumeration

### PR DESCRIPTION
**Pull request by wangdi1**
Since VOS aggregation might merge the recxs during recx enumeration,
so let's add logical anchor probe, to make sure the anchor can always
be probed correctly.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

File: src/vos/evt_iter.c [hunk 1]
```diff
@@ -109,6 +109,30 @@ evt_iter_finish(daos_handle_t ih)
 	return rc;
 }
 
+static int
+evt_rect_sorted_cmp(const struct evt_rect *rt1, const struct evt_rect *rt2)
```
Comment (jolivier23) [inline src/vos/evt_iter.c:113]: Add a comment and change the name
/** For anchor matching, we want to match the rectangle if it's at the same epoch and
 *.  there is overlap.   This ensures we see extents that may get merged into an extent 
  *  we've already seen.
  */
evt_rect_anchor_cmp

I think using sorted is confusing since evt_rect_cmp is sorted.

File: src/vos/evt_iter.c [hunk 2]
```diff
@@ -358,7 +382,16 @@ evt_iter_probe_sorted(struct evt_context *tcx, struct evt_iterator *iter,
 	index = evt_iter_probe_find(iter, rect);
 	if (index == -1)
 		return -DER_NONEXIST;
+
 	iter->it_index = index;
+	entry = evt_ent_array_get(iter->it_entries, index);
+	if (iter->it_forward)
+		entry->en_sel_ext.ex_lo = max(rect->rc_ex.ex_lo, entry->en_ext.ex_lo);
```
Comment (jolivier23) [inline src/vos/evt_iter.c:389]: This seems weird.   Why isn't it already fixed up when selecting the rectangle to begin with?

Also, can the low and high both be wrong (e.g. one rectangle in the tree with smaller selection ?)

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/vos/evt_iter.c (hunk 1 above)
Comment (wangdi1) [reply src/vos/evt_iter.c:113]: sure

File: src/vos/evt_iter.c (hunk 2 above)
Comment (wangdi1) [reply src/vos/evt_iter.c:389]: > This seems weird. Why isn't it already fixed up when selecting the rectangle to begin with?
> 

Hmm, I thought evt_iter_probe_sorted is selected the started rectangle for iteration.  So we either fix the started rectangle here in this allocated array, or fix during iter_fetch,  but this seems clearer and simpler here. Or I miss the question?

> Also, can the low and high both be wrong (e.g. one rectangle in the tree with smaller selection ?)

hmm, I thought evt_iter_probe_find() will make sure the index is located either on the overlapped extent or the extent closet with the rect anchor.   And this update will make sure the following iteration starting  

1. if no overlap extent is found, then start from the closet extent in the array.
2. if there is overlap extent,  then start from the anchor extent.

**Comment by wangdi1** [comment]
ping, this is 2.0.1 blocker.

**Comment by johannlombardi** [comment]
please inspect.

**Review by jolivier23 — DISMISSED** [review]

File: src/vos/evt_iter.c (hunk 2 above)
Comment (NiuYawei) [reply src/vos/evt_iter.c:394]: I don't see why it's adjusted as "max(anchor.ex_lo, recx.ex_lo)", is it possible that "recx.ex_lo > anchor.ex_lo"? If so, you'll partially (or fully) skip anchor, if recx.ex_lo is always greater than anchor.ex_lo, you'd just put an assert and use anchor.ex_lo here.

Is this iterator probe used only for rebuild? Or you want it being used by DAOS recx enumeration API as well? If it'll be used by DAOS recx enumeration API as well, I have few questions:

1. Will you change the recx enumeration to use sorted evtree iterator in follow-on patch?
2. What's the semantic of recx enumeration API will be? I mean what's the expected behavior if the anchor is merged with new updates, or even fully deleted by aggregation? How will application deal with such situation? I think you'd clarify all these in the commit comment.

I personally think that we didn't clearly define the recx enumeration API, if we still want to support (and improve it), we'd clearly define the expected behavior so that application can have a definite expectation.

File: src/vos/evt_iter.c [hunk 3]
```diff
@@ -358,7 +387,16 @@ evt_iter_probe_sorted(struct evt_context *tcx, struct evt_iterator *iter,
...
+	else
+		entry->en_sel_ext.ex_hi = min(rect->rc_ex.ex_hi, entry->en_ext.ex_hi);
```
Comment (NiuYawei) [inline src/vos/evt_iter.c:396]: Same question, is it possible that "recx.ex_hi < anchor.ex_hi"?

**Review by NiuYawei — CHANGES_REQUESTED** [review]

File: src/vos/evt_iter.c [hunk 4]
```diff
@@ -109,6 +109,35 @@ evt_iter_finish(daos_handle_t ih)
 	return rc;
 }
 
+/**
+ * For anchor matching, we want to match the rectangle if it's at the same epoch and
+ * there is overlap. This ensures we see extents that may get merged into an extent
+ * we've already seen.
+ */
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/vos/evt_iter.c:116]: >Let's take a simple example, assuming anchor is [2, 3)@epoch 10, the merged recx is [1, 4)@epoch 20, will this compare function return '>' or '<'?

I thought aggregation will choose lower epoch?  so if [2, 3) is merged to [1, 4), its epoch should not be higher than 10?

So let's say merging extent is [1, 4) @epoch 5,  let's see this list

[0, 1)@epoch2, [1, 4) @epoch5,  [5, 7)@epoch6,  [8, 10)@epoch8  

then after binary search of [2, 3)@epoch10  in evt_iter_probe_find(), it will return mid=1 ([1, 4)epoch5.

>then how can you match the anchor?

So it does not matter whether there is matched record in this list,  but the mid (return by evt_iter_probe_find() have to be the closest one, where it is left.

**Comment by wangdi1** [comment]
>is it possible that "recx.ex_lo > anchor.ex_lo"? If so, you'll partially (or fully) skip anchor, if recx.ex_lo is always greater than anchor.ex_lo, you'd just put an assert and use anchor.ex_lo here.

oh, it is impossible for recx.ex_lo > anchor.ex_lo by evt_iter_probe_find(), unless the anchor recx has been deleted, in that case, this max() will make sure it will start from next recx.


>Is this iterator probe used only for rebuild? Or you want it being used by DAOS recx enumeration API as well? If it'll be used by DAOS recx enumeration API as well.

Hmm, this is not specifically for rebuild, IMHO, we need this for normal recx enumeration as well, given VOS aggregation might happen any time. Please check the test I added here.


>Will you change the recx enumeration to use sorted evtree iterator in follow-on patch?

Yes, normal recx enumeration will use this anyway, see my test case here.

>What's the semantic of recx enumeration API will be? I mean what's the expected behavior if the anchor is merged with new updates, or even fully deleted by aggregation? How will application deal with such situation? I think you'd clarify all these in the commit comment.

I think this is very general case, i.e. it just trying to locate the recx anchor, which is an range,  instead of trying to match the exact recx anchor, that is the only difference here. TBH, I think the comments borrowed from @jolivier23  said this clearly.  Or I miss sth here?
Hmm, even it is deleted, evt_probe will make sure it will start from the "right" recx.

File: src/vos/evt_iter.c (hunk 3 above)
Comment (wangdi1) [reply src/vos/evt_iter.c:396]: >Same question, is it possible that "recx.ex_hi < anchor.ex_hi"?

unless the anchor recx is deleted, otherwise it is impossible.

**Comment by wangdi1** [comment]
Update the patch to fix the reverse iteration issue, please check.

File: src/vos/evt_iter.c (hunk 3 above)
Comment (jolivier23) [reply src/vos/evt_iter.c:384]: Rather than fixing it here, it would be better, I think to fix the filter in evt_iter_probe_sorted.

Set the upper or lower bound of the range in the filter according to what is in the anchor.  That way, we don't have to do this here as it will already be set.   And if we are half way through the array, we can filter out and avoid sorting half of the rectangles.

So the variable you want to change is iter->it_filter

**Review by jolivier23 — CHANGES_REQUESTED** [review]

**Review by jolivier23 — DISMISSED** [review]

**Comment by kprantis** [comment]
Does this need to merge master to pickup the fix that caused not all testing to run?

**Comment by wangdi1** [comment]
> Does this need to merge master to pickup the fix that caused not all testing to run?

yes, sure.

**Comment by kprantis** [comment]
Tests passed, but looks like there are comments still on the patch?

**Comment by wangdi1** [comment]
Tests passed, but looks like there are comments still on the patch?

I think we already resolved all of comments during Skype discussion.

**Comment by kprantis** [comment]
Ok, so, we need Niu to re-review and/or clear his comments if he's already re-reviewed, then ready for gatekeeping?

**Comment by wangdi1** [comment]
> Ok, so, we need Niu to re-review and/or clear his comments if he's already re-reviewed, then ready for gatekeeping?

Yes, we need another approve here, then ask for landing.

File: src/vos/evt_iter.c [hunk 5]
```diff
@@ -154,16 +172,15 @@ evt_iter_probe_find(struct evt_iterator *iter, const struct evt_rect *rect)
 		return mid;
 
 	/* We didn't find the entry.  Find next entry instead */
+	/* NB: it does not need adjust for reverse iteration, because
+	 * mid <= anchor after the above while.
+	 */
```
Comment (NiuYawei) [inline src/vos/evt_iter.c:177]: This looks incorrect to me, when reverse iterator failed to find anchor, it should the next smaller one. Could you revert this change?

Comment (wangdi1) [reply src/vos/evt_iter.c:177]: hmm, I think we do not need adjust the mid for reverse find, since the mid is already on the smaller side if it can not find.  Though we probably need check for start == end case.

**Review by NiuYawei — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Comment by johannlombardi** [comment]
Are failures known?

**Comment by kprantis** [comment]
At least 3 runs had test failures, so just a note there are multiple runs to look at if failures are known.

**Comment by wangdi1** [comment]
@johannlombardi @kprantis [  ](https://github.com/daos-stack/daos/pull/7881)  based on 2.0.1 release actually passed all tests,  could we land that one? if this is still blocker for 2.0.1.  Thanks.

**Comment by kprantis** [comment]
#22 didn't actually run any tests. I was tryign to re-run from #21, and didn't notice it was selected on the bullseye report rather than the HW stage, so all it did was rerun the bullseye stage and not re-run the HW tests as I intended.

By 3 runs, I mean, there are failures in #21, #23, and #24, all of which we need to evaluate to ensure they are known issues and not an introduced regression.

**Comment by wangdi1** [comment]
> #22 didn't actually run any tests. I was tryign to re-run from #21, and didn't notice it was selected on the bullseye report rather than the HW stage, so all it did was rerun the bullseye stage and not re-run the HW tests as I intended.
> 
> By 3 runs, I mean, there are failures in #21, #23, and #24, all of which we need to evaluate to ensure they are known issues and not an introduced regression.

Ah, I see. Sorry, misunderstood here.

**Comment by wangdi1** [comment]
#21 failure is actually due to a segfault.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-7759/21/artifact/Functional%20Hardware%20Medium/daos_test/suite.py/daos_logs/DAOS_Management_daos_control.wolf-139.log/*view*/    
Saw this in quite a few other test already.  let me create a ticket.

**Comment by kprantis** [comment]
Even more failures to disposition in #25 I am afraid.

**Comment by wangdi1** [comment]
> Even more failures to disposition in #25 I am afraid.

Sigh, most of them are environment failure. 

 Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_engine[191900]: accel_engine.c: 988:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_engine[191900]: EAL: Cannot find device (0000:81:00.0)
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_engine[191900]: EAL: Failed to attach device on primary process
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_engine[191900]: nvme.c: 813:nvme_probe_internal: *ERROR*: NVMe ctrlr scan failed
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]: ERROR: daos_engine:0 EAL: Cannot find device (0000:81:00.0)
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]: EAL: Failed to attach device on primary process
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_engine[191900]: bdev_nvme.c:2691:bdev_nvme_create: *ERROR*: No controller was found with provided trid (traddr: 0000:81:00.0)
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]: ERROR: daos_engine:0 [2022-01-18 20:55:24.378058] nvme.c: 813:nvme_probe_internal: *ERROR*: NVMe ctrlr scan failed
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]: [2022-01-18 20:55:24.378075] bdev_nvme.c:2691:bdev_nvme_create: *ERROR*: No controller was found with provided trid (traddr: 0000:81:00.0)
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_engine[191900]: json_config.c: 220:rpc_client_poller: *ERROR*: error response: 
                                                                      {
                                                                        "code": -19,
                                                                        "message": "No such device"
                                                                      }
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]: ERROR: daos_engine:0 [2022-01-18 20:55:24.381308] json_config.c: 220:rpc_client_poller: *ERROR*: error response:
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]: {
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]:   "code": -19,
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]:   "message": "No such device"
    Jan 18 20:55:24 wolf-202.wolf.hpdd.intel.com daos_server[189271]: }

**Comment by wangdi1** [comment]
> Test stage Functional Hardware Large completed with status FAILURE. https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-7759/26/display/redirect

The new failure is just some known DTX failure here. sigh.
